### PR TITLE
feature/#37-named-by-sheet

### LINKED ISSUE #37: readNamedRegion and Local (sheet scoped) Names ?
I have two sheets in my workbook. In each one  a local (sheet-scoped) name "Hugo" is defined refering to the cell B2 in the specific worksheet. The values in the cells are 122 resp. 222.

> getDefinedNames(wb)
> [1] "Hugo" "Hugo"
> 
> readNamedRegion(wb, 'Hugo')
> [1] Col1
> <0 rows> (or 0-length row.names)

From my point of view it seems to be unpossible to retrieve the value 222 stored in the cell named "Hugo" in Worksheet(2) via readNamedRegion.

To reproduce the issue the following Excel Macro might be helpful:
Sub DefineNameHugoInTwoSheets()
  Dim wsX As Worksheet, i As Integer, sSheetName As String, sNameName As String
  For i = 1 To 2
    Set wsX = Worksheets(i)
    wsX.Activate
    sSheetName = wsX.Name
    sNameName = "Hugo"
    wsX.Names.Add sNameName, "=$B$2"
    wsX.Range(sNameName).Value = i \* 100 + 22
  Next i
End Sub

### PR DESCRIPTION
**Pull request by spoltier**
Closes #37 - see miraisolutions/xlconnect-java#10

Sorry about the bad diffs, it was a while ago and my rstudio setup was probably not quite right

### CODE REVIEW

**Comment by spoltier** [comment]
Still have to complete documentation

**Comment by spoltier** [comment]
Discussion - need to check scope of name definition (where is the name valid) vs. sheet where the referenced data is defined; are all the functions consistent with both of these ?

**Comment by riccardoporreca** [comment]
@spoltier, regarding the worksheet scope vs reference, I guess we can confirm the target is to introduce support for worksheet-scoped names.

Besides some specific technical / consistency feedback (coming soon), here some general inputs from my side

- The argument is currently called `worksheetName`. Shall we be more explicit about its meaning by calling it `worksheetScope`? Otherwise, if we don't need to be specific w.r.t. to the argument being the name scope, we can perhaps just use `sheet`, which is what we use as argument in most of the existing XLConnect functions requiring a worksheet.

- I would recommend defining the default argument of the user-exposed R function `NULL` instead of `.jnull(class = "java/lang/String")`, which is more natural and can better support programming around the XLConnect functions.
  - We can have an internal utility resolving R's `NULL` to Java's within each function at the time of calling the Java interface; perhaps even defining the quite common infix operator `%||%` and using it inline as `worksheetName %||% .jnull(class = "java/lang/String")`, see e.g. in [rlang](https://github.com/r-lib/rlang/blob/2a0dddb8f184d23ed1aa120be7c00b1216ed5de4/R/operators.R#L1-L16).

- With reading data from a named region in mind, I think we need to support three cases
  - A) Same behavior as before this change if no / NULL argument specified, also for backwards compatibility: the "first" defined name is used.
  - B) Explicit sheet-scoped name, which is pretty much what is implemented now with the new argument.
  - C) Explicit workbook-scoped name: this is **missing**, meaning we do not provide any way to read a workbook-scoped name explicitly, unless it is (by chance?) the name retrieved as the "first" in case (A). => We could implement this as `worksheetScope = FALSE`: How to best propagate this to the Java side with a `String` argument is of course a question, perhaps just `""` converted to sheet index -1 (global) internally.

File: R/workbook.readNamedRegion.R [hunk 1]
```diff
@@ -41,20 +41,27 @@ setMethod("readNamedRegion",
...
 
 		# returns a list of RDataFrameWrapper Java object references
-		sheet = as.vector(extractSheetName(getReferenceFormula(object, name)))
-		namedim = matrix(as.vector(t(getReferenceCoordinatesForName(object, name))), nrow=4, byrow=FALSE)
+	  if(identical(worksheetName,.jnull(class = "java/lang/String"))) {
+		  sheet = as.vector(extractSheetName(getReferenceFormula(object, name)))
+	  }
+	  else {
+	    sheet = as.vector(worksheetName)
+	  }
```
Comment (riccardoporreca) [inline R/workbook.readNamedRegion.R:52]: The `sheet` variable defined here is use to target subsets of columns in the data via `getColSubset()`. As such it should always be the sheet where the data are, regardless of the scope where the name is defined., 

I guess we need to extend `getReferenceFormula()` to gain the worksheet scope argument and just use it as

```r
sheet = as.vector(extractSheetName(getReferenceFormula(object, name, worksheetName)))
```

File: R/writeNamedRegionToFile.R [hunk 2]
```diff
@@ -1,51 +1,52 @@
...
+# Wrapper function to write data to an Excel file in one line.
+# 
+# Author: Thomas Themel, Mirai Solutions GmbH
+#
+#############################################################################
+
+
+writeNamedRegionToFile <- function(file, data, name, formula = NA, ..., 
+		styleAction = XLC$STYLE_ACTION.XLCONNECT,clearNamedRegions=FALSE) {
+
```
Comment (riccardoporreca) [inline R/writeNamedRegionToFile.R:32]: We need to make use of the possible `worksheetName` argument (retrieved as e.g. `list(...)$worksheetName` or added to the function signature explicitly):

- When handling `clearNamedRegions` (might require quite some refactoring)
- Passed to `createName()`

**Comment by riccardoporreca** [comment]
Besides the comments to specific modified function, we also miss extending `removeName()` with the additional argument.

Perhaps it might be good to extend `getDedinedNames()` to have `worksheetName` argument, allowing to retrieve all, sheet-scoped, workbook-scoped names. This might be useful also for handling `clearNamedRegions` in the comment above https://github.com/miraisolutions/xlconnect/pull/146#discussion_r712833373.

File: R/writeNamedRegionToFile.R (hunk 2 above)
Comment (spoltier) [reply R/writeNamedRegionToFile.R:32]: @riccardoporreca I've made changes to enable this, but part of it seems like a hack (use of `sapply`). 

Without it, the following occurs, which I don't really understand:
```
Error in mapply(FUN = function (...)  : 
  zero-length inputs cannot be mixed with those of non-zero length
> traceback()
12: mapply(FUN = function (...) 
    .jrcall(x, name, ...), character(0), list(new("jobjRef", jobj = <pointer: (nil)>, 
        jclass = "java/lang/Object")), SIMPLIFY = FALSE)
11: do.call("mapply", args = c(FUN = f, args, SIMPLIFY = .simplify))
10: doTryCatch(return(expr), name, parentenv, handler)
9: tryCatchOne(expr, names, parentenv, handlers[[1L]])
8: tryCatchList(expr, classes, parentenv, handlers)
7: tryCatch(..., Throwable = function(e) {
       if (!is.jnull(e$jobj)) {
           sw <- new(J("java.io.StringWriter"))
           pw <- new(J("java.io.PrintWriter"), sw)
           e$jobj$printStackTrace(pw)
           options(java.stacktrace = sw$toString())
           stop(paste(class(e)[1], e$jobj$getMessage(), sep = " (Java): "), 
               call. = FALSE)
       }
       else stop("Undefined error occurred!")
   })
6: jTryCatch(do.call("mapply", args = c(FUN = f, args, SIMPLIFY = .simplify)))
5: xlcCall(object, "clearNamedRegion", name, worksheetScope %||% 
       .jnull(), .simplify = FALSE)
4: clearNamedRegion(wb, toClear, worksheetScope = worksheetScope)
3: clearNamedRegion(wb, toClear, worksheetScope = worksheetScope)
2: writeNamedRegionToFile(file, df, name, formula = paste(worksheet, 
       "A1", sep = "!"))
```

Comment (riccardoporreca) [reply R/writeNamedRegionToFile.R:32]: @spoltier, need to look at this more in detail and give it some more thoughts

File: R/xlcAttributesCall.R [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      colnames(attrMtx) <- allANames
+      
+      Map(function(aName) {
+        attr(unwrapped, aName) <- attrMtx[, aName]
+      }, allANames)
+    }
+    else
+      attr(unwrapped, allANames[1]) <- attrMtx
+    unwrapped
+  }
```
[the remark this answers is not in the document]
Comment (spoltier) [reply R/xlcAttributesCall.R:80]: Current status: attributes now show up for readNamedregion.
To do next:
[ ] implement toggle for disabling setting attributes by default
[ ] add tests checking attribute functionality 
[ ] Implement for further relevant calls including getDefinedNames
[ ] document attribute functionality
...

**Pull request closed without merging**